### PR TITLE
Better Thread Safety

### LINKED ISSUE #54: Crashes on [RZDBObserver invalidate]
Hi, thank you for a great library, I really enjoy using it.

Recently, I've been receiving tens of crashes on ```[RZDBObserver invalidate]``` with an exception ```*** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[__NSArrayM objectAtIndex:]: index 189 beyond bounds [0 .. 188]'``` (this reason keeps repeating, but with different numbers like ```index 605 beyond bounds [0 .. 604]' ```.

```
Last Exception Backtrace:
0   CoreFoundation                       0x27a29c1f 0x27924000 + 1072159
1   libobjc.A.dylib                      0x35210c8b 0x3520a000 + 27787
2   CoreFoundation                       0x2793e839 0x27924000 + 108601
3   CoreFoundation                       0x27956b77 0x27924000 + 207735
4   RZDataBinding                        0x00fbd0fd -[RZDBObserver invalidate] (NSObject+RZDataBinding.m:391)
5   CoreFoundation                       0x279512df 0x27924000 + 185055
6   CoreFoundation                       0x27951209 0x27924000 + 184841
7   RZDataBinding                        0x00fbc9ad -[NSObject(RZDataBinding_Private) rz_cleanupObservers] (NSObject+RZDataBinding.m:308)
8   RZDataBinding                        0x00fbd799 __rz_swizzleDeallocIfNeeded_block_invoke_2 (NSObject+RZDataBinding.m:496)
9   libobjc.A.dylib                      0x3522ad5f 0x3520a000 + 134495
10  CoreFoundation                       0x27930e5d 0x27924000 + 52829
11  CoreFoundation                       0x279444e7 0x27924000 + 132327
12  libobjc.A.dylib                      0x3522ad5f 0x3520a000 + 134495
```

Do you have any idea where the problem could be?

Thanks!

### PR DESCRIPTION
**Pull request by jvisenti**
- Switched to using `RZDBObserverContainer` in place of `NSMutableArray`
- Use `object_getAssociatedObject`/`object_setAssociatedObject` directly instead of method wrappers
- Added thread safe enumeration to `RZDBObserverContainer`

Attempts to fix #54

### CODE REVIEW

File: RZDataBinding/NSObject+RZDataBinding.m [hunk 1]
```diff
@@ -409,28 +388,62 @@ - (void)invalidate
...
+}
+
+- (void)enumerateObserversUsingBlock:(void (^)(RZDBObserver *, BOOL *))block
+{
+    @synchronized (self) {
+        NSHashTable *observers = ^{
+            @synchronized (_observers) {
+                return [_observers copy];
+            }
+        }();
```
Comment (KingOfBrian) [inline RZDataBinding/NSObject+RZDataBinding.m]: What is this, swift?!

I was looking at replacing it with the C-based `({ })` format, but this still reads nicer.

File: RZDataBinding/NSObject+RZDataBinding.m [hunk 2]
```diff
@@ -195,52 +201,32 @@ - (void)rz_unbindKey:(NSString *)key fromKeyPath:(NSString *)foreignKeyPath ofOb
...
 
         if ( registeredObservers == nil ) {
-            registeredObservers = [NSMutableArray array];
-            [self rz_setRegisteredObservers:registeredObservers];
+            registeredObservers = [RZDBObserverContainer strongContainer];
+            rz_setRegisteredObservers(self, registeredObservers);
         }
 
-        [registeredObservers addObject:observer];
+        [registeredObservers addObserver:observer];
```
Comment (KingOfBrian) [inline RZDataBinding/NSObject+RZDataBinding.m]: You can move this line out of the `@synchronized(self)`

File: RZDataBinding/NSObject+RZDataBinding.m [hunk 3]
```diff
@@ -195,52 +201,32 @@ - (void)rz_unbindKey:(NSString *)key fromKeyPath:(NSString *)foreignKeyPath ofOb
...
+        dependentObservers = rz_dependentObservers(target);
 
         if ( dependentObservers == nil ) {
-            dependentObservers = [[RZDBObserverContainer alloc] init];
-            [target rz_setDependentObservers:dependentObservers];
+            dependentObservers = [RZDBObserverContainer weakContainer];
+            rz_setDependentObservers(target, dependentObservers);
         }
 
         [dependentObservers addObserver:observer];
```
Comment (KingOfBrian) [inline RZDataBinding/NSObject+RZDataBinding.m]: Same here

**Comment by KingOfBrian** [comment]
Looks good @jvisenti -- The observer container reads nicely and minimizes the amount of exposed `@synchronized` nicely.

File: RZDataBinding/NSObject+RZDataBinding.m (hunk 1 above)
Comment (jvisenti) [inline RZDataBinding/NSObject+RZDataBinding.m]: Oh yeah, I knew there was another syntax, but I couldn't remember haha. So change this or no?

File: RZDataBinding/NSObject+RZDataBinding.m (hunk 2 above)
Comment (jvisenti) [inline RZDataBinding/NSObject+RZDataBinding.m]: Actually, I don't think the `@synchronized` is needed anymore

File: RZDataBinding/NSObject+RZDataBinding.m (hunk 3 above)
Comment (jvisenti) [inline RZDataBinding/NSObject+RZDataBinding.m]: Same here, `@synchronized` is no longer necessary

File: RZDataBinding/NSObject+RZDataBinding.m (hunk 2 above)
Comment (jvisenti) [inline RZDataBinding/NSObject+RZDataBinding.m]: Hmm interestingly the actual KVO observer is added above here. Think there needs to be some sync after all, and I'm not sure if it was even right before

**Pull request merged**